Userland/Libraries/LibHTTP/Job.cpp

```diff
@@ -82,45 +82,122 @@ Job::~Job()
 {
 }
 
+void Job::start(Core::Stream::Socket& socket)
+{
+    VERIFY(!m_socket);
+    m_socket = static_cast<Core::Stream::BufferedSocketBase*>(&socket);
+    dbgln_if(HTTPJOB_DEBUG, "Reusing previous connection for {}", url());
+    deferred_invoke([this] {
+        dbgln_if(HTTPJOB_DEBUG, "HttpJob: on_connected callback");
+        on_socket_connected();
+    });
+}
+
+void Job::shutdown(ShutdownMode mode)
+{
+    if (!m_socket)
+        return;
+    if (mode == ShutdownMode::CloseSocket) {
+        m_socket->close();
+    } else {
+        m_socket->on_ready_to_read = nullptr;
+        m_socket = nullptr;
+    }
+}
+
 void Job::flush_received_buffers()
 {
     if (!m_can_stream_response || m_buffered_size == 0)
         return;
     dbgln_if(JOB_DEBUG, "Job: Flushing received buffers: have {} bytes in {} buffers for {}", m_buffered_size, m_received_buffers.size(), m_request.url());
-    for (size_t i = 0; i < m_received_buffers.size(); ++i) {
-        auto& payload = m_received_buffers[i];
-        auto written = do_write(payload);
-        m_buffered_size -= written;
-        if (written == payload.size()) {
-            // FIXME: Make this a take-first-friendly object?
-            m_received_buffers.take_first();
-            --i;
-            continue;
+    for (size_t try_ = 0; try_ < 8; ++try_) {
+        for (size_t i = 0; i < m_received_buffers.size(); ++i) {
+            auto& payload = m_received_buffers[i];
+            auto result = do_write(payload);
+            if (result.is_error()) {
+                if (!result.error().is_errno()) {
+                    dbgln_if(JOB_DEBUG, "Job: Failed to flush received buffers: {}", result.error());
+                    continue;
+                }
+                if (result.error().code() == EINTR) {
+                    i--;
+                    continue;
+                }
+                break;
+            }
```

Comment:
This pattern is repeated more than once now, should `PosixSocketHelper` check for `EINTR` itself to simplify this code?

Reply:
Would be nice.